### ISSUE TITLE
Responce Headers

### ISSUE THREAD

**Issue by tvh** [haskell-servant/servant#25]
Currently it is impossible to specify in any way which headers are returned. It would be nice if there was a way to specify this. Related issues: https://github.com/haskell-servant/servant-server/issues/25 https://github.com/haskell-servant/servant-client/issues/16.

As far as I can tell a `ResponseHeader X a` similar to `Header` should be implementable by collecting them in a type-level list on the way. I am however unsure about what the response type should be.

**Comment by christian-marie** [comment]
I agree that this seems implementable with a ResponseHeader '["MAGIC1", "MAGIC2"] kind of thing.

**Comment by alpmestan** [comment]
Do we only lack a way to set response headers or are there other things we might want to set in the response? (we already have the response body -- obviously -- and the content-type)

The thing is, it's a bit hard to compose response-related combinators because we conceptually only have one "slot" to describe the response: the last element in a `:>`-separated chain of combinators, which necessarily has to be one of `Get`, `Put`, `Post`, `Delete` or `Patch`. So how would you see it?

``` haskell
-- * first approach
type Api1 = "u" :> Header "X-Preferred-Haskell-Library" Text :> ResponseHeader "X-Preferred-Haskell-Library" Text (Get [JSON] User)
-- we stack headers by "composing" the combinator with itself
type Api1' = ResponseHeader "a" Text (ResponseHeader "b" Text (Get [JSON] User))

-- * second approach
type Api2 = "u" :> Header "X-Preferred-Haskell-Library" Text :> ResponseHeaders [("X-Preferred-Haskell-Library", Text)] (Get [JSON] User)
-- all response headers are specified in the same place

-- * any other approaches in mind?
```

And how would the combinator affect the handler type? `EitherT (Int, String) IO a` becoming `EitherT (Int, String) IO (a, Text)`?

**Comment by jkarni** [comment]
A modified content-type approach should work:

```
type Api a = SetsHeader '[ETag, Location] :> Get '[JSON] a
```
Would check that there is a class, `C`, s.t. `C ETag a` and `C Location a`, which defines a method `mkHeader :: a -> (ByteString, ByteString)`. That way, you don't have to repeatedly tuple anything, and can still use whatever type you want.

**Comment by alpmestan** [comment]
We might want to add another argument somewhere, a "response thats being built up", and have the instances modify it? So that we can tweak the response as we go and keep the combinatorial approach for response-generation.

**Comment by christian-marie** [comment]
The content-type approach is precisely what I had in mind when I conjectured that it should be implementable.

Another argument (that's passed through the instances) of "a response being built up" would work, also. I wonder if we can generalize that to just "state" that's being built up so that we don't feel like we are special casing something every time we want to add a new thing like this.

**Comment by christian-marie** [comment]
We could even run within a state monad and bind "backwards" through the instances, if that's not too weird.

**Comment by tvh** [comment]
@christian-marie When you say content-type, do you mean just extend the MimeRender and MimeUnrender classes to include the headers?

**Comment by christian-marie** [comment]
I was confused, apologies. I really just meant a type level list approach.

**Comment by tvh** [comment]
I don't really like the idea of polluting the constructors further. Extending the existing classes could work like

```.haskell
type SimpleApi = Get '[Header '["FooHeader"] JSON] Bar

class Accept ctype => MimeUnrender ctype a where
    fromByteString :: Proxy ctype -> ByteString -> [(ByteString, ByteString)] -> Either String a

class Accept ctype => MimeRender ctype a where
    toByteString  :: Proxy ctype -> a -> (ByteString, [(ByteString, ByteString)])
```

**Comment by christian-marie** [comment]
Looks a little weird from the user's perspective. I can see that working, though.

**Comment by jkarni** [comment]
@tvh but this is really confusing - it seems to mix Accept/Content-Type specific logic (which requires parsing the Accept header in a very specific way) with other headers.

@alpmestan I think your idea of threading state could solve a number of problems - possibly even the fact that InvalidBody gets thrown before InvalidMethod (and maybe the performance issue with `Canonicalize` that Sönke pointed out). 

But it's a pretty significant change to my mind, so I'm inclined to say that we shouldn't do it for 0.3.

**Comment by christian-marie** [comment]
@tvh Can you drop a link to a use case? Like the OpenStack client you're building?

**Comment by thsutton** [comment]
The changes to `MimeRender` and `MimeUnrender` seem really weird to me. I wouldn't expect, for example, that rendering a `Foo` as `JSON` might return different HTTP headers than when rending it as `XML`.

Yet that's exactly what this API change is saying.

**Comment by tvh** [comment]
@thsutton Yes, that is indeed weird... Another alternative with this approach would be to introduce a new typeClass `ResponseRender` and use that insteaad.

@jkarni Threading state through would be the ideal solution, yes. When I tried that however, I couldn't figure out how to make headers optional without introducing yet another copy of the `Get`, `Delete`, ... instances and even more overlapping instances. Maybe this change would even be possible as a pure addition to existing stiff, but I am not sure about that. If so, it would certainly be the preferred option.

@christian-marie The specific use-case is the OpenStack-Keystone-client, yes. https://github.com/anchor/openstack-api/blob/master/lib/OpenStack/Keystone.hs

**Comment by alpmestan** [comment]
Solved with #28
